### PR TITLE
Major refactor of kube-aws tool

### PR DESCRIPTION
**Pull request by colhom**
* kube-aws can now generate all assets/templates. We can now distribute just the binary.

* No S3 intermediary. All assets are compressed on `kube-aws up` and sent directly to cloudformation api.

* `kube-aws render` generates a reproducible cluster asset pack which contains tls assets and all templates

* Much simpler templating pipeline. Single templating phase on `kube-aws up`.

* New `validate` command to validate user-data and stack json

* Replace confusing cloudformation parameters `kube-aws up --update`, which will update the cloudformation stack.

* Unit tests for all config modules

* Replace `godeps` with `glide`!

### CODE REVIEW

**Comment by colhom** [comment]
Pretty much everything has changed. To try it out for yourself, check out the [new readme](https://github.com/coreos/coreos-kubernetes/tree/repro-assets-2/multi-node/aws)

We aren't hosting the pre-compiled binary yet, so you'll have to compile it yourself. Once you have the binary, there is no dependency on the project tree. The design goal here is to be able to distribute "just the binary" when we release.

**Comment by colhom** [comment]
Here's the working list of things that should be sorted out before we roll this up into a major release. The goal is to offer a "production ready solution" for provisioning a coreos kubernetes cluster. These are the major functionality blockers that I can think of.

* [ ] Use AWS KMS to secure TLS assets. Provision/specify KMS key, and have kube-aws encrypt TLS assets, and nodes decrypt them using the API. The cloudformation will no longer contain un-encrypted TLS secrets! (Shout out to @mfburnett for this idea)

* [ ] Support multiple controller/etcd instances (HA control plane)

* [ ] EBS volume attachment for controller(s). Update procedure will detach EBS volume from old controller instance and reattach to new controller instance. This will preserve cluster state in the event the controller(s) are updated and replaced as a result of cloudformation update.

* [ ] Secure etcd peer/client connections with TLS
 
* [ ] Route53 integration for APIServerEndpoint. Automatically create hosted-zone and/or A record for controller EIP on `kube-aws up`

* [ ] Put controller(s) into AutoscalingGroup

* [ ] Set up controller and worker AutoscalingGroups to recover from ec2 instance failures

**Comment by aaronlevy** [comment]
Would you be able to break out the godeps updates into a separate commit(s) to make reviewing a bit easier?

**Comment by colhom** [comment]
great idea.

**Comment by colhom** [comment]
@aaronlevy I've addressed the "isolated" issues on the PR so far with this batch of commits. @ericchiang is going to help figure out a better templating/buffer approach for this PR.

He requested that we review these commits, squash it up and break out 3 discrete commits
* Go Code
* Vendored Deps
* Docs

Then the discussion can continue

**Comment by colhom** [comment]
@ericchiang added some commits building on what you've got. the tool works now and is feature complete with where it was.

cc\ @aaronlevy

**Comment by ericchiang** [comment]
@colhom it might be best to squash these commits now. They're a little unreadable ATM (I know I'm partially responsible for this).

**Comment by colhom** [comment]
@aaronlevy, commits split up and rebased on master.

File: .travis.yml [hunk 1]
```diff
@@ -1,12 +1,15 @@
 language: go
 go:
-  - 1.4
   - 1.5
 install:
+  - export GO15VENDOREXPERIMENT=1
+  - wget https://github.com/Masterminds/glide/releases/download/0.8.3/glide-0.8.3-linux-amd64.tar.gz
```
Comment (aaronlevy) [inline .travis.yml]: Are you still using glide?

File: .travis.yml [hunk 2]
```diff
@@ -1,12 +1,15 @@
...
+  - tar zxvf glide-0.8.3-linux-amd64.tar.gz
+  - export PATH="${PATH}:$(pwd)/linux-amd64"
   - go get golang.org/x/tools/cmd/vet
 script:
   - cd $TRAVIS_BUILD_DIR/multi-node/aws
-  - GOPATH=${PWD}/Godeps/_workspace:${GOPATH} go test -v ./...
-  - test -z "$(find . -path ./Godeps -prune -o -name '*.go' -exec gofmt -d {} + | tee /dev/stderr)"
-  - go vet ./...
+  - test -z "$(find . -prune -o -name '*.go' -exec gofmt -d {} + | tee /dev/stderr)"
```
Comment (aaronlevy) [inline .travis.yml]: Is the `-prune -o` necessary anymore now that `./Godeps` is no longer available? Do we want to prune the vendor directory instead?

File: .travis.yml [hunk 3]
```diff
@@ -1,12 +1,15 @@
...
   - ./build
+  - go test -v $(find ./{cmd,pkg} -maxdepth 1 -mindepth 1 -type d)
+  - go vet $(find ./{cmd,pkg} -maxdepth 1 -mindepth 1 -type d)
```
Comment (aaronlevy) [inline .travis.yml]: Curious why these need to be changed from `./...` ?

File: multi-node/aws/README.md [hunk 4]
```diff
@@ -5,24 +5,86 @@ View the full instructions at https://coreos.com/kubernetes/docs/latest/kubernet
 
 ## Development
 
+### Download pre-built binary
+
+```sh
+wget https://<binary-url>
```
Comment (aaronlevy) [inline multi-node/aws/README.md]: If this gets merged, this should not be a placeholder

File: multi-node/aws/README.md [hunk 5]
```diff
@@ -5,24 +5,86 @@ View the full instructions at https://coreos.com/kubernetes/docs/latest/kubernet
...
+# check checksum
+chmod +x ./kube-aws
+sudo mv kube-aws /usr/bin/
```
Comment (aaronlevy) [inline multi-node/aws/README.md]: I would leave this up to the user to decide to move into path or not. Nothing wrong with just using this tool outside path.

File: multi-node/aws/README.md [hunk 6]
```diff
@@ -5,24 +5,86 @@ View the full instructions at https://coreos.com/kubernetes/docs/latest/kubernet
...
+* golang >= 1.5
+
 The compiled binary will be available at `./bin/kube-aws`.
 
-### Custom Kubernetes Manifests
+## Initialize an asset directory
+```sh
+$ mkdir my-cluster
+$ cd ./my-cluster
+$ kube-aws init --cluster-name=my-cluster-name --external-dns-name=my-cluster-endpoint --region=us-west-1 --availability-zone=us-west-1c --key-name=key-pair-name
```
Comment (aaronlevy) [inline multi-node/aws/README.md]: Similar to comment above, if you were able to point to a target directory, the bin being available on the path isn't as necessary E.g. `./kube-aws init --cluster-dir=~/clusters/foobar`

File: multi-node/aws/cmd/kube-aws/command_destroy.go [hunk 7]
```diff
@@ -17,32 +17,29 @@ var (
...
-	err := cluster.DecodeConfigFromFile(cfg, rootOpts.ConfigPath)
+	cfg, err := config.NewConfigFromFile(configPath)
 	if err != nil {
-		stderr("Unable to load cluster config: %v", err)
+		stderr("Error parsing config: %v", err)
 		os.Exit(1)
 	}
 
-	c := cluster.New(cfg, newAWSConfig(cfg))
+	cluster := cluster.New(cfg, destroyOpts.awsDebug)
```
Comment (aaronlevy) [inline multi-node/aws/cmd/kube-aws/command_destroy.go]: var name conflicts with package name

File: multi-node/aws/build [hunk 8]
```diff
@@ -14,16 +14,17 @@ if [ -n "$(git diff --shortstat 2> /dev/null | tail -n1)" ]; then
     VERSION="${VERSION}+dirty"
 fi
 
-export GO15VENDOREXPERIMENT=1
+export GOPATH="$(pwd)/_output"
+export KUBEAWS="$GOPATH/src/github.com/coreos/coreos-kubernetes/multi-node/aws"
 
+if [ -d $GOAPTH ]; then
```
Comment (aaronlevy) [inline multi-node/aws/build]: GOPATH

File: multi-node/aws/build [hunk 9]
```diff
@@ -14,16 +14,17 @@ if [ -n "$(git diff --shortstat 2> /dev/null | tail -n1)" ]; then
...
+    rm -rf $GOPATH
```
Comment (aaronlevy) [inline multi-node/aws/build]: nit: the if/then directory check around this is redundant with -rf

File: multi-node/aws/build [hunk 10]
```diff
@@ -14,16 +14,17 @@ if [ -n "$(git diff --shortstat 2> /dev/null | tail -n1)" ]; then
...
-mkdir -p $link_path
-ln -s ${PWD}/../../ "${link_path}/coreos-kubernetes"
-
-GOPATH=${PWD}/Godeps/_workspace:${PWD}/gopath CGO_ENABLED=0 go build -ldflags "-X main.VERSION ${VERSION}" -a -tags netgo -installsuffix netgo -o bin/kube-aws github.com/coreos/coreos-kubernetes/multi-node/aws/cmd/kube-aws
-rm -r ./gopath
+mkdir -p $GOPATH/src
+cp -r ./vendor/* $GOPATH/src
+mkdir -p $KUBEAWS
+cp -r ./pkg $KUBEAWS/pkg
+cp -r ./cmd $KUBEAWS/cmd
```
Comment (aaronlevy) [inline multi-node/aws/build]: Could be more lightweight with symlinks

File: multi-node/aws/pkg/cluster/cluster.go [hunk 11]
```diff
@@ -64,181 +58,61 @@ func (c *Cluster) stackName() string {
...
+	if report != "" {
+		fmt.Printf("Update stack: %s\n", report)
 	}
-
-	tmplURL := fmt.Sprintf("%s/template.json", c.cfg.ArtifactURL)
-	return createStackAndWait(cloudformation.New(c.aws), c.stackName(), tmplURL, parameters)
+	return err
 }
 
+//TODO: validate cluster
```
Comment (aaronlevy) [inline multi-node/aws/pkg/cluster/cluster.go]: nit: add name to TODOs (and in this case provide more information)

**Comment by aaronlevy** [comment]
Should we be committing `multi-node/aws/pkg/config/templates.go` or can we have that generated during the build process?

**Comment by aaronlevy** [comment]
Ugh and this is saying all my comments are out of date :( 

Is there any reason to keep the original commit if it's basically thrown away? I'll start commenting on the last commit .. but this is difficult to review

File: multi-node/aws/pkg/cluster/cluster.go [hunk 12]
```diff
@@ -44,83 +42,146 @@ func New(cfg *config.Config, awsDebug bool) *Cluster {
...
+			time.Sleep(3 * time.Second)
+			continue
+		default:
+			return "", fmt.Errorf("unexpected stack status: %s", statusString)
+		}
 	}
-	return err
 }
 
 //TODO: validate cluster
```
Comment (aaronlevy) [inline multi-node/aws/pkg/cluster/cluster.go]: nit: add name to TODOs (also more info would be helpful)

**Comment by aaronlevy** [comment]
Closing in favor of https://github.com/coreos/coreos-kubernetes/pull/315.

Same code, just cleaned up a bit to make review easier.

**Comment by aaronlevy** [comment]
I lied, this was moved to: https://github.com/coreos/coreos-kubernetes/pull/316

**Pull request closed without merging**